ignition-server/controllers/tokensecret_controller.go

```diff
@@ -113,40 +111,30 @@ func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	if err := r.Client.Get(ctx, req.NamespacedName, tokenSecret); err != nil {
 		if apierrors.IsNotFound(err) {
 			log.Info("Token Secret not found", "request", req.String())
+			// If a secret was deleted, ensure that any corresponding tokens for that
+			// secret are purged from the cache.
+			for _, k := range r.PayloadStore.Keys() {
+				v, ok := r.PayloadStore.Get(k)
+				if ok && v.SecretName == req.Name {
+					log.Info("Deleting token for missing secret", "tokenSecret", req.Name)
+					r.PayloadStore.Delete(k)
+				}
+			}
 			return ctrl.Result{}, nil
 		}
 		return ctrl.Result{}, err
 	}
 
-	// If the tokenSecret is not being deleted then ensure the finalizer.
-	if tokenSecret.DeletionTimestamp.IsZero() {
-		if !controllerutil.ContainsFinalizer(tokenSecret, finalizer) {
-			controllerutil.AddFinalizer(tokenSecret, finalizer)
-			if err := r.Update(ctx, tokenSecret); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
-			}
-		}
-	}
-
-	// If the tokenSecret is being deleted then delete from cache and remove finalizer.
-	if !tokenSecret.DeletionTimestamp.IsZero() {
```

Comment:
Shouldn't we keep this in case the cache does return the secret after it was deleted?

Reply:
It couldn't hurt. I'll add it back.

Reply:
I went ahead and incorporated this edge case into the PR.